src/Functions/namedTupleItems.cpp

```diff
@@ -0,0 +1,130 @@
+#include <Functions/IFunction.h>
+#include <Functions/FunctionFactory.h>
+#include <Functions/FunctionHelpers.h>
+#include <DataTypes/IDataType.h>
+#include <DataTypes/DataTypeTuple.h>
+#include <DataTypes/DataTypeArray.h>
+#include <DataTypes/DataTypeString.h>
+#include <Columns/ColumnTuple.h>
+#include <Columns/ColumnArray.h>
+#include <Columns/ColumnString.h>
+#include <Columns/ColumnsNumber.h>
+#include <Common/assert_cast.h>
+#include <memory>
+
+namespace DB
+{
+namespace ErrorCodes
+{
+    extern const int ILLEGAL_TYPE_OF_ARGUMENT;
+    extern const int ILLEGAL_INDEX;
+}
+
+namespace
+{
+
+/** Extract element of tuple by constant index or name. The operation is essentially free.
+  * Also the function looks through Arrays: you can get Array of tuple elements from Array of Tuples.
+  */
+class FunctionNamedTupleItems : public IFunction
+{
+public:
+    static constexpr auto name = "namedTupleItems";
+    static FunctionPtr create(ContextPtr)
+    {
+        return std::make_shared<FunctionNamedTupleItems>();
+    }
+
+    String getName() const override
+    {
+        return name;
+    }
+
+    size_t getNumberOfArguments() const override
+    {
+        return 1;
+    }
+
+    bool useDefaultImplementationForConstants() const override
+    {
+        return true;
+    }
+
+    DataTypePtr getReturnTypeImpl(const ColumnsWithTypeAndName & arguments) const override
+    {
+        // get the type of all the fields in the tuple
+        const IDataType * col = arguments[0].type.get();
+        const DataTypeTuple * tuple = checkAndGetDataType<DataTypeTuple>(col);
+
+        if (!tuple)
+            throw Exception("First argument for function " + getName() + " must "
```

Comment:
Please consider to use fmt like syntax.
```
throw Exception(ErrorCodes::ILLEGAL_TYPE_OF_ARGUMENT, "First argument for function {} must be a tuple.", getName());
```

Reply:
Ah I didn't know there was such a constructor, that's good.